### PR TITLE
[PERF] make internal model lazier

### PR DESCRIPTION
**Pull request by runspired**
Branches off of heimdall-instrumentation, so that should be merged fist ;)

### CODE REVIEW

**Comment by fivetanley** [comment]
so excited :pray:

we'll have to make sure we don't break @bookingsync's and others' addons.

**Comment by runspired** [comment]
@fivetanley this provides identical private API's to what was present before, they are just lazy, with one notable exception. I've removed the use of `get` and `set` for `currentState`. It was being used inconsistently, half the places used dot notation, half used get/set. Looking into the code paths it appears there isn't a strong case for it needing to be `get/set` based, and obvs with it being inconsistent before if it did "need to be" it would have been broken.

**Comment by fivetanley** [comment]
yeah i am not really concerned about the api changing here, just a last step before merging to run it against addons that use `internalModel` heavily.

**Comment by runspired** [comment]
Also note, the tests on this pass locally but I need to rebase off master for them to pass here, If we could merge heimdall today I've got a follow up PR after this one that's also fun :)

**Comment by runspired** [comment]
@igor rebase complete

File: addon/-private/system/model/internal-model.js [hunk 1]
```diff
@@ -851,26 +946,26 @@ InternalModel.prototype = {
...
       }
     }
 
     return changedKeys;
-  },
+  }
 
   toString() {
     if (this.record) {
       return this.record.toString();
```
Comment (stefanpenner) [inline addon/-private/system/model/internal-model.js]: this branch seems un-needed

File: addon/-private/system/model/states.js [hunk 2]
```diff
@@ -584,6 +584,7 @@ var RootState = {
       },
 
       didCommit(internalModel) {
+        // TODO @runspired lastDirtyType doesn't seem to be a real thing
         internalModel.send('invokeLifecycleCallbacks', get(internalModel, 'lastDirtyType'));
```
Comment (stefanpenner) [inline addon/-private/system/model/states.js]: ^ todo + get, if its not used then this line may be a candidate for removal

File: addon/-private/system/model/internal-model.js [hunk 3]
```diff
@@ -574,37 +570,49 @@ export default class InternalModel {
...
       state = state.parentState;
-    } while (!state.hasOwnProperty(pivotName));
+    } while (!state[pivotName]);
+
+    let setups;
+    let enters;
+    let i;
+    let l;
+
+    if (TransitionChainMap[transitionMapId]) {
```
Comment (stefanpenner) [inline addon/-private/system/model/internal-model.js]: maybe `let map = TransitionChainMap[transitionMapId]` then reference map everywhere

File: addon/-private/system/model/states.js [hunk 4]
```diff
@@ -241,7 +244,7 @@ var DirtyState = {
 
     //TODO(Igor) reloading now triggers a
     //loadingData event, though it seems fine?
-    loadingData: Ember.K,
+    loadingData: K,
```
Comment (stefanpenner) [inline addon/-private/system/model/states.js]: lets just do `loadingData() { }` rather then the single shared function of `K`

File: addon/-private/system/model/states.js [hunk 5]
```diff
@@ -4,7 +4,10 @@
 import Ember from 'ember';
 import { assert } from "ember-data/-private/debug";
 
-var get = Ember.get;
+const {
+  K
```
Comment (stefanpenner) [inline addon/-private/system/model/states.js]: lets not use `K` regular concise functions are likely fine.

**Comment by runspired** [comment]
Eyeballing the production build benchmarks (no filtering of outliers but I am looking at a box plot), we moved from roughly 35ms spent in `store.push` for 238 records of a single type to a bit less than 28ms.

**Comment by tchak** [comment]
LGTM 👍

File: addon/-private/system/model/internal-model.js [hunk 6]
```diff
@@ -16,11 +16,22 @@ import {
...
+  isEmpty,
+  isEqual,
+  run: emberRun,
+  RSVP,
+  RSVP: { Promise }
+} = Ember;
+
+const assign = Ember.assign || Ember.merge;
+const setOwner = Ember.setOwner;
+const TransitionChainMap = new EmptyObject();
```
Comment (igorT) [inline addon/-private/system/model/internal-model.js]: Can you put a comment here or in the method as to how the transitionchain map works

File: addon/-private/system/model/model.js [hunk 7]
```diff
@@ -305,6 +308,7 @@ var Model = Ember.Object.extend(Ember.Evented, {
     @private
     @type {Object}
   */
+  currentState: null,
```
Comment (stefanpenner) [inline addon/-private/system/model/model.js]: should we define this on the prototype

File: addon/-private/system/model/states.js [hunk 8]
```diff
@@ -430,7 +433,7 @@ createdState.uncommitted.pushedData = function(internalModel) {
   internalModel.triggerLater('didLoad');
 };
 
-createdState.uncommitted.propertyWasReset = Ember.K;
+createdState.uncommitted.propertyWasReset = K;
```
Comment (stefanpenner) [inline addon/-private/system/model/states.js]: `= function() { } `

**Comment by runspired** [comment]
@igorT Travis is passing after making the changes from @stefanpenner's last review.

**Pull request merged**